### DISCUSSION TITLE
Lilygo TTGO TSIM7000G GSM example

### DISCUSSION THREAD

**Discussion by tyeth** [micropython/micropython#9399] [General]
In answer to my own enquiries and the closed(archived) discussion here https://forum.micropython.org/viewtopic.php?f=18&t=8248

This is an example of basic GSM connectivity for SIM7000 on ESP32 
https://github.com/Xinyuan-LilyGO/LilyGO-T-SIM7000G/blob/master/examples/MicroPython_LoBo/gsm.py

**Comment by rwslippey** [comment]
Sorry to hijack your thread here but I've been having a heck of a time getting this moving at all. 

I started off with the same example code you referenced above but with warnings that ultimately stopped execution. I did use the recommended version of micropython.

I actually also built https://github.com/Xinyuan-LilyGO/lilygo-micropython but there appears to be no real documentation on how to work with the build for the GSM or GPS modules, the code above also does not work on their build.... 

Thanks for any ideas or even just a direction to some info.

**Reply by tyeth** [reply]
So there's some peeps trying to get it moving using esp_modem module, an official esp-idf module, not really had a need myself yet apart from prove it connected. 

What were the errors/warnings you eventually saw (after how long?), and did you then try the reset procedure which I read a little about? Also curious if you got it recently or ages ago and whether you attempted a firmware update on the SIM7000G?

It's worth reading about the other modules, I've got the t-PCI-E which has plug in modules, for which I have the 7070G and 7600G. There are more answers for those/other SIM7xxxx repos/boards and often similar/identical initialisation procedure (2G/GPRS has a slightly different command set but other modules using 3g/LTE seemed to share almost everything). The lilygo person was incredibly helpful and responsive to questions that have been already asked. I learned a lot when I browsed closed issues in the lilygo repos and some of the tinyGPS repo and  adafruit FONA code which I think was the Sim800 (2G/GPRS so not exactly same as later modules).

**Reply by rwslippey** [reply]
> So there's some peeps trying to get it moving using esp_modem module, an official esp-idf module, not really had a need myself yet apart from prove it connected.
> 
> What were the errors/warnings you eventually saw (after how long?), and did you then try the reset procedure which I read a little about? Also curious if you got it recently or ages ago and whether you attempted a firmware update on the SIM7000G?
> 
> It's worth reading about the other modules, I've got the t-PCI-E which has plug in modules, for which I have the 7070G and 7600G. There are more answers for those/other SIM7xxxx repos/boards and often similar/identical initialisation procedure (2G/GPRS has a slightly different command set but other modules using 3g/LTE seemed to share almost everything). The lilygo person was incredibly helpful and responsive to questions that have been already asked. I learned a lot when I browsed closed issues in the lilygo repos and some of the tinyGPS repo and adafruit FONA code which I think was the Sim800 (2G/GPRS so not exactly same as later modules).

Quite honestly, Ive been down so many paths the last two days, I've lost track of the errors. So kind of the "mile high overview" is. 
I started with the recommended version of micropython from https://github.com/Xinyuan-LilyGO/LilyGO-T-SIM7000G/tree/master/examples/MicroPython_LoBo and the example code there. I can't recall the errors there, but now I can't even flash it... so I'll work that end here... Using any other version of micropython results in import errors on the gsm module.

Seeing that the previous examples were kind of old and it looks like LoBo hasn't been very active, I pivoted to trying to build based on https://github.com/Xinyuan-LilyGO/lilygo-micropython, which I successfully built, and works fine, except I can't find any documentation at all for how to actually use it.

I'll take a look at your suggestions, I think really the issue is just lack of documentation. I'm fairly confident in python, so micropython is a nature move but I'm pretty new here. I've never been very good at Arduino and I wasn't able to make much work there either. I was able to test AT commands but beyond that here I am

**Comment by rwslippey** [comment]
Finally a bit of success here. After reading tyeth's reply I started poking around in the lilygo-miropython repo and found /extmod/network/at/test/. Looking around the code helps to understand a bit about the modules that appear to be included in the build generated from the repo. So that helps. Still having troubles with GPRS but GPS works!

**Reply by tyeth** [reply]
I found running through the commands from the T-Sim7000 arduino sketch example works fine, and as far as I remember I maybe tried the same first few AT commands in Python REPL as the arduino sketch does to test for available network/gps, but I only tested an MQTT example in arduino for example not python. The AT command code I was using for testing in python may have included some commands out of a github issue (maybe from another SIM7070/7800 repo), not very helpful sorry, that was the hope of this discussion page - to collate info :)

**Reply by rwslippey** [reply]
Ha, no worries...  I'm basically just firing in the dark here... I did manage to get GPRS connection though. So far I've managed to send a TCP ping to my provider, first I've actually seen on hologram's dashboard that I'm on the network, so that works. Also managed to actually get an IP now battling the SHCONN +CME Error: Operation now allowed deal... it's been a battle but I'm making progress and wrapping my head around things.   Good news is I'm probably 90% of the way there. My purpose here is a stupid simple so I'm just gonna wrap up my testing into a few functions and it'll be good, everything else is done on my server so, just need to get to the api!  full speed ahead

**Comment by kjm1102** [comment]
Sure would be nice to use urequests or similar with the 7000G. I've spent ages getting the 7080G to download ~19k .mpy files using that horrendous AT+SHCHEAD system that waveshare advocate but it's ugly. A nice lib that would work with vanilla upython would make a life a lot easier.

**Comment by JayPalm** [comment]
@rwslippey Curious where you've gotten with this? Hoping to switch from the LoBo fork and just found this thread. Were you able to get a webREPL via GSM? Still building the firmware so haven't been able to test anything.

**Comment by rwslippey** [comment]
@JayPalm I got a bit distracted with other projects recently. I did manage to get a few things working... although in a hacky way.

First, I was able to kind of build a mess that gets me a connection and I could hit my api server. I don't currently have it in a git repo but I can drop it in one here this week. It's ugly though....  Mostly this was building AT commands manually based on the manual.

After I got that working, I actually found in the repository https://github.com/Xinyuan-LilyGO/lilygo-micropython there is some included libs (in the libs directory, who would have thought...) that offered a class but most of them are not complete and show "TODO" so it appears maybe that was abandoned and needs some work from the community to get those classes setup. I was thinking of doing this and got pulled away.

I'm by no means a pro programmer but I did write a small article to remind myself of where I'm at. I hate self promotion but it might be helpful to you...  https://www.robslippey.com/lilygo-t-sim7000g-gps-and-cellular-data-in-micropython/

I can circle back to this in a bit and drop a repo online with what I've got once I sanitize my API creds. 

Also, sorry for the delay, I never got a notification on the mention

**Reply by JayPalm** [reply]
@rwslippey all good. Been banging my head against this for a couple days and just finally making some progress tonight to establish a GPRS connection and send/receive messages via MQTT.

I've also discovered the [LilyGo repo](https://github.com/Xinyuan-LilyGO/lilygo-micropython) and spent some time trying to make a build (I seemed to be having some issues related to using an M1, so I offloaded the build to my VPS and it went fine). I also saw the libGSM.c and gsm_PPP.c libraries in the LilyGo repo, but as far as I could tell, they're not actually included in the build as configured, because the make file has a conditional that's never set to true. I tried removing the conditional and rebuilding, which did give me a gsm module importable in the REPL, but I couldn't figure out what the arguments were in the function call were. I also started to implement some of those modem level AT commands in the classes that they started and left to wither on the vine, will push at some point hopefully this weekend.

Anyway, what actually worked I think didn't involve any of that. The official micropython esp32 port has a network.PPP module. Here's a [gist](https://gist.github.com/JayPalm/deed52e6b8f7b394409679e1aebd0c41) of what I did to actually get things working. I am using the custom built firmware I mentioned above, but I'm pretty sure this would work on firmware built from the main project, though I'll try to test that this weekend as well.

The modem does seem to be a bit unsteady I've noticed and seems to shut itself off sometimes, then resist being turned back on. The module is powered over USB, so I wouldn't think the power supply is an issue but maybe it's a crappy cable. Anyway, more work to be done this weekend.

I appreciate your write up. No problem with self promotion. I just can't understand why it's taken you and me both so many hours to get this thing up and running. It's like some kind of forbidden knowledge!

**Reply by carTloyal123** [reply]
I have not had the chance to try any of this yet but I am appreciate the work going on here and hopefully can contribute in the future! I bought some of LilyGo SIM7000G as a cheap way to deploy data collectors and would love to use the internet connection cleanly without AT commands and this is definitely the most convincing information I have seen yet so thank you!

**Reply by tyeth** [reply]
There was some heavy info in a github issue about resetting when the modem enters an unstable/error state, something about the procedure was very specific, and the reset time was surprisingly large, also get the commands or procedure out of order and the modem is unusable until the correct procedure followed. This issue wasn't the one I read and was thinking of, but covers the same ground: https://github.com/Xinyuan-LilyGO/LilyGO-T-SIM7000G/issues/53
